### ISSUE TITLE
False Negatives in Unit Tests

### ISSUE THREAD

**Issue by clanmills** [Exiv2/exiv2#763]
The following are cosmetic.  However, it's probably a small effort to hide these distractions.

1) On Cygwin and MinGW:
```
[----------] 3 tests from strError
[ RUN      ] strError.returnSuccessAfterClosingFile
/home/rmills/gnu/github/exiv2/buildserver/unitTests/test_futils.cpp:31: Failure
      Expected: expectedString
      Which is: "Success (errno = 0)"
To be equal to: strError().c_str()
      Which is: "No error (errno = 0)"
[  FAILED  ] strError.returnSuccessAfterClosingFile (2 ms)
[ RUN      ] strError.returnNoSuchFileOrDirectoryWhenTryingToOpenNonExistingFile
[       OK ] strError.returnNoSuchFileOrDirectoryWhenTryingToOpenNonExistingFile (0 ms)
[ RUN      ] strError.doNotRecognizeUnknownError
[       OK ] strError.doNotRecognizeUnknownError (0 ms)
[----------] 3 tests from strError (2 ms total)
```
I'm not sure of the origin of strings 'Success' or 'No Error', however we can evade this by only report `if error != 0`.

2) On all platforms, we're getting date/time warnings, from tests which correctly warn as follows:
```Warning: Unsupported time format```

For example (on Linux)
```
[----------] 9 tests from ADateValue
[ RUN      ] ADateValue.isDefaultConstructed
[       OK ] ADateValue.isDefaultConstructed (0 ms)
...
[ RUN      ] ADateValue.doNotReadFromByteBufferWithoutExpectedSize
Warning: Unsupported date format
....
```

The flag `--quiet` on the exiv2 command-line program will suppress warnings and we should consider using that on those tests.

**Comment by D4N** [comment]
Robin Mills <notifications@github.com> writes:

> The following are cosmetic.  However, it's probably a small effort to hide these distractions.
>
> 1) On Cygwin and MinGW:
> ```
> [----------] 3 tests from strError
> [ RUN      ] strError.returnSuccessAfterClosingFile
> /home/rmills/gnu/github/exiv2/buildserver/unitTests/test_futils.cpp:31: Failure
>       Expected: expectedString
>       Which is: "Success (errno = 0)"
> To be equal to: strError().c_str()
>       Which is: "No error (errno = 0)"
> [  FAILED  ] strError.returnSuccessAfterClosingFile (2 ms)
> [ RUN      ] strError.returnNoSuchFileOrDirectoryWhenTryingToOpenNonExistingFile
> [       OK ] strError.returnNoSuchFileOrDirectoryWhenTryingToOpenNonExistingFile (0 ms)
> [ RUN      ] strError.doNotRecognizeUnknownError
> [       OK ] strError.doNotRecognizeUnknownError (0 ms)
> [----------] 3 tests from strError (2 ms total)
> ```
> I'm not sure of the origin of strings 'Success' or 'No Error', however we can evade this by only report `if error != 0`.

Yes, we should do that. The code in that unit test is checking
`strerror(errno)`, but we already have something like 4 or 5 #ifdefs
there for all the various platforms, because they all have different
error strings. Adding a sixth isn't going to make this better.

>
> 2) On all platforms, we're getting date/time warnings, from tests which correctly warn as follows:
> ```Warning: Unsupported time format```
>
> For example (on Linux)
> ```
> [----------] 9 tests from ADateValue
> [ RUN      ] ADateValue.isDefaultConstructed
> [       OK ] ADateValue.isDefaultConstructed (0 ms)
> ...
> [ RUN      ] ADateValue.doNotReadFromByteBufferWithoutExpectedSize
> Warning: Unsupported date format
> ....
> ```
>
> The flag `--quiet` on the exiv2 command-line program will suppress warnings and we should consider using that on those tests.
>

The unit tests never invoke the exiv2 application directly, so that
isn't going to help. But I guess the `--quiet` options sets some global
flag, so we could set that in a test setup function (and reset it in
tearDown again).

**Comment by piponazo** [comment]
I noticed that we are not running the unit tests in the Fedora_MinGW Gitlab CI job. I think it would be good to enable it at some point and fix any possible issue we might find there.

For the first case you mentioned, it is mostly a problem of the test code not being complete. I wrote those tests as "Characterisation tests" since I wanted to know more about how these functions were behaving. Interestingly, `strError` (and internally `strerror_r`) are returning different values for different systems. 

Now that we are talking about it ... it feels like those tests are testing `strerror_r` rather than our `strError` method ... I think I could change them slightly to overcome this situation.

**Comment by clanmills** [comment]
I believe this was fixed on 'master' in #802, however it doesn't appear to have been fixed on 0.27-maintenance.

**Comment by piponazo** [comment]
I am afraid this is not easily applicable in 0.27. We would need to make considerable changes in the CI scripts to install gmock there (I gave it a try, but I think does not worth it the effort). I propose to skip this change from 0.27 since it is only an issue in the testing code, but not in production code (and now we are mostly taking care about security issues in 0.27).
